src/wp-includes/rest-api/class-wp-rest-server.php

```diff
@@ -884,6 +908,73 @@ public function envelope_response( $response, $embed ) {
 		return rest_ensure_response( $envelope );
 	}
 
+	/**
+	 * Register a namespace for lazy loading
+	 *
+	 * @param string $route_namespace The namespace to register for lazy loading
+	 */
+	public function register_lazy_loaded_namespace( $route_namespace ) {
+		if ( ! isset( $this->lazy_namespaces[ $route_namespace ] ) ) {
+			$this->lazy_namespaces[ $route_namespace ] = false;
+		}
+	}
+
+	/**
+	 * Load a specific lazy namespace
+	 *
+	 * @param string $route_namespace
+	 *
+	 * @return bool True if loaded, false if not a lazy namespace
+	 */
+	protected function load_lazy_namespace( $route_namespace ) {
+		if ( ! isset( $this->lazy_namespaces[ $route_namespace ] ) ) {
+			return false;
+		}
+
+		if ( $this->lazy_namespaces[ $route_namespace ] ) {
+			return true; // Already loaded
+		}
+
+		$this->lazy_namespaces[ $route_namespace ] = true;
+
+		/**
+		 * Fires when a lazy-loaded REST API namespace is being loaded.
+		 *
+		 * This action is triggered when a lazy-loaded namespace needs to have its
+		 * routes registered. Plugins should hook into this action to register
+		 * their routes for the specific namespace being loaded.
+		 *
+		 * The dynamic portion of the hook name, `$route_namespace`, refers to
+		 * the namespace being loaded (e.g., 'my-plugin/v1').
+		 *
+		 * @since X.X.0
+		 *
+		 * @example
+		 * // Register routes when your namespace loads
+		 * add_action( 'rest_lazy_load_namespace_my-plugin/v1', function() {
+		 *     register_rest_route( 'my-plugin/v1', '/posts', array(
+		 *         'methods'             => 'GET',
+		 *         'callback'            => 'my_plugin_get_posts',
+		 *         'permission_callback' => '__return_true',
+		 *     ) );
+		 * } );
+		 */
+		do_action( 'rest_lazy_load_namespace_' . $route_namespace );
```

Comment:
This somewhat breaks from convention around only using underscores for hooks since by their own convention namespaces are almost always going to contain dashes and slashes.  Would it be better convert these for the hook for the sake of convention or leave it as is?

Reply:
This also isn't a recommended pattern anymore. The namespace should just be it's own parameter.